File: frontend/src/lib/mock/data/records.ts
import type { AerospikeRecord, GeoJSON, BinValue } from "@/lib/api/types";

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

const firstNames = [
  "Alice", "Bob", "Charlie", "Diana", "Eve", "Frank", "Grace", "Hank",
  "Irene", "Jack", "Karen", "Leo", "Mona", "Nick", "Olivia", "Paul",
  "Quinn", "Rita", "Sam", "Tina", "Uma", "Vic", "Wendy", "Xander",
  "Yara", "Zane",
];

const lastNames = [
  "Smith", "Johnson", "Williams", "Brown", "Jones", "Garcia", "Miller",
  "Davis", "Rodriguez", "Martinez", "Hernandez", "Lopez", "Gonzalez",
  "Wilson", "Anderson", "Thomas", "Taylor", "Moore", "Jackson", "Martin",
];

const productNames = [
  "Wireless Mouse", "Mechanical Keyboard", "USB-C Hub", "Monitor Stand",
  "Webcam HD", "Desk Lamp", "Laptop Stand", "Cable Organizer",
  "Noise-Cancelling Headphones", "Bluetooth Speaker", "External SSD",
  "Portable Charger", "Smart Watch", "Fitness Tracker", "VR Headset",
  "Drone Mini", "Action Camera", "E-Reader", "Tablet Stylus", "Docking Station",
  "Ring Light", "Microphone USB", "Game Controller", "Mouse Pad XL",
  "Screen Protector", "Phone Mount", "Car Charger", "Wall Adapter",
  "HDMI Cable", "Ethernet Adapter",
];

const categories = [
  "Electronics", "Accessories", "Audio", "Computing", "Gaming",
  "Photography", "Wearables", "Storage",
];

const orderStatuses = [
  "pending", "processing", "shipped", "delivered", "cancelled", "refunded",
];

const streetNames = [
  "Main St", "Oak Ave", "Cedar Ln", "Elm Dr", "Pine Rd",
  "Maple Ct", "Birch Blvd", "Walnut Way", "Spruce Pl", "Ash Cir",
];

const cities = [
  "San Francisco", "New York", "Chicago", "Austin", "Seattle",
  "Denver", "Boston", "Portland", "Miami", "Atlanta",
];

const states = [
  "CA", "NY", "IL", "TX", "WA", "CO", "MA", "OR", "FL", "GA",
];

const tags = [
  "premium", "new", "verified", "vip", "beta-tester",
  "newsletter", "referral", "loyalty", "early-adopter", "enterprise",
];

function seededRandom(seed: number): () => number {
  let state = seed;
  return () => {
    state = (state * 1664525 + 1013904223) & 0x7fffffff;
    return state / 0x7fffffff;
  };
}

function pick<T>(arr: T[], rand: () => number): T {
  return arr[Math.floor(rand() * arr.length)];
}

function pickN<T>(arr: T[], n: number, rand: () => number): T[] {
  const shuffled = [...arr].sort(() => rand() - 0.5);
  return shuffled.slice(0, n);
}

function generateDigest(rand: () => number): string {
  const hex = "0123456789abcdef";
  let digest = "";
  for (let i = 0; i < 40; i++) {
    digest += hex[Math.floor(rand() * hex.length)];
  }
  return digest;
}

// ---------------------------------------------------------------------------
// User records (conn-1, test:users)
// ---------------------------------------------------------------------------

function generateUserRecords(): AerospikeRecord[] {
  const rand = seededRandom(42);
  const records: AerospikeRecord[] = [];

  for (let i = 1; i <= 50; i++) {
    const firstName = pick(firstNames, rand);
    const lastName = pick(lastNames, rand);
    const email = `${firstName.toLowerCase()}.${lastName.toLowerCase()}${i}@example.com`;
    const age = 18 + Math.floor(rand() * 52);
    const active = rand() > 0.2;
    const userTags = pickN(tags, 1 + Math.floor(rand() * 4), rand);

    const metadata: Record<string, BinValue> = {
      signupDate: `2025-${String(1 + Math.floor(rand() * 12)).padStart(2, "0")}-${String(1 + Math.floor(rand() * 28)).padStart(2, "0")}`,
      lastLogin: `2026-02-${String(1 + Math.floor(rand() * 14)).padStart(2, "0")}T${String(Math.floor(rand() * 24)).padStart(2, "0")}:${String(Math.floor(rand() * 60)).padStart(2, "0")}:00Z`,
      loginCount: Math.floor(rand() * 500),
      preferredLanguage: pick(["en", "es", "fr", "de", "ja"], rand),
      timezone: pick(["US/Pacific", "US/Eastern", "UTC", "Europe/London", "Asia/Tokyo"], rand),
    };

    records.push({
      key: {
        namespace: "test",
        set: "users",
        pk: `user-${String(i).padStart(4, "0")}`,
        digest: generateDigest(rand),
      },
      meta: {
        generation: 1 + Math.floor(rand() * 10),
        ttl: -1,
        lastUpdateMs: Date.now() - Math.floor(rand() * 86400000 * 30),
      },
      bins: {
        name: `${firstName} ${lastName}`,
        email,
        age,
        active,
        metadata,
        tags: userTags,
      },
    });
  }
  return records;
}

// ---------------------------------------------------------------------------
// Product records (conn-1, test:products)
// ---------------------------------------------------------------------------

function generateProductRecords(): AerospikeRecord[] {
  const rand = seededRandom(137);
  const records: AerospikeRecord[] = [];

  for (let i = 1; i <= 30; i++) {
    const name = productNames[(i - 1) % productNames.length];
    const price = Math.round((5 + rand() * 495) * 100) / 100;
    const category = pick(categories, rand);
    const inStock = rand() > 0.15;

    const specs: Record<string, BinValue> = {
      weight: `${Math.round(50 + rand() * 2000)}g`,
      dimensions: `${Math.round(5 + rand() * 30)}x${Math.round(5 + rand() * 20)}x${Math.round(2 + rand() * 10)}cm`,
      color: pick(["Black", "White", "Silver", "Blue", "Red", "Gray"], rand),
      warranty: pick(["6 months", "1 year", "2 years", "3 years"], rand),
      manufacturer: pick(["TechCorp", "GadgetPro", "DigitalWorks", "NexGen", "InnoTech"], rand),
      rating: Math.round((3 + rand() * 2) * 10) / 10,
      reviewCount: Math.floor(rand() * 1200),
    };

    records.push({
      key: {
        namespace: "test",
        set: "products",
        pk: `prod-${String(i).padStart(4, "0")}`,
        digest: generateDigest(rand),
      },
      meta: {
        generation: 1 + Math.floor(rand() * 5),
        ttl: -1,
        lastUpdateMs: Date.now() - Math.floor(rand() * 86400000 * 60),
      },
      bins: {
        name,
        price,
        category,
        inStock,
        specs,
      },
    });
  }
  return records;
}

// ---------------------------------------------------------------------------
// Order records (conn-2, staging:orders)
// ---------------------------------------------------------------------------

function generateOrderRecords(): AerospikeRecord[] {
  const rand = seededRandom(256);
  const records: AerospikeRecord[] = [];

  for (let i = 1; i <= 200; i++) {
    const orderId = `ORD-${String(10000 + i)}`;
    const userId = `user-${String(1 + Math.floor(rand() * 50)).padStart(4, "0")}`;
    const status = pick(orderStatuses, rand);

    const itemCount = 1 + Math.floor(rand() * 5);
    const items: BinValue[] = [];
    let total = 0;
    for (let j = 0; j < itemCount; j++) {
      const itemPrice = Math.round((10 + rand() * 200) * 100) / 100;
      const qty = 1 + Math.floor(rand() * 3);
      total += itemPrice * qty;
      items.push({
        productId: `prod-${String(1 + Math.floor(rand() * 30)).padStart(4, "0")}`,
        name: pick(productNames, rand),
        price: itemPrice,
        quantity: qty,
      });
    }
    total = Math.round(total * 100) / 100;

    const stateIdx = Math.floor(rand() * states.length);
    const address: Record<string, BinValue> = {
      street: `${100 + Math.floor(rand() * 9900)} ${pick(streetNames, rand)}`,
      city: cities[stateIdx],
      state: states[stateIdx],
      zip: String(10000 + Math.floor(rand() * 89999)),
      country: "US",
    };

    // Realistic US coordinates: lat ~25-48, lng ~(-124)-(-71)
    const lat = Math.round((25 + rand() * 23) * 1000000) / 1000000;
    const lng = Math.round((-124 + rand() * 53) * 1000000) / 1000000;
    const location: GeoJSON = {
      type: "Point",
      coordinates: [lng, lat],
    };

    records.push({
      key: {
        namespace: "staging",
        set: "orders",
        pk: orderId,
        digest: generateDigest(rand),
      },
      meta: {
        generation: 1 + Math.floor(rand() * 3),
        ttl: 86400 * 90, // 90 days TTL
        lastUpdateMs: Date.now() - Math.floor(rand() * 86400000 * 30),
      },
      bins: {
        orderId,
        userId,
        total,
        status,
        items,
        address,
        location,
      },
    });
  }
  return records;
}

// ---------------------------------------------------------------------------
// Export
// ---------------------------------------------------------------------------

/**
 * Mock records keyed by connection ID, then by "namespace:set".
 */
export const mockRecords: Record<string, Record<string, AerospikeRecord[]>> = {
  "conn-1": {
    "test:users": generateUserRecords(),
    "test:products": generateProductRecords(),
  },
  "conn-2": {
    "staging:orders": generateOrderRecords(),
  },
};

File: backend/src/aerospike_ui_api/mock_data/records.py
from __future__ import annotations

import math
import time

from aerospike_ui_api.models.record import AerospikeRecord, RecordKey, RecordMeta

# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------

FIRST_NAMES = [
    "Alice",
    "Bob",
    "Charlie",
    "Diana",
    "Eve",
    "Frank",
    "Grace",
    "Hank",
    "Irene",
    "Jack",
    "Karen",
    "Leo",
    "Mona",
    "Nick",
    "Olivia",
    "Paul",
    "Quinn",
    "Rita",
    "Sam",
    "Tina",
    "Uma",
    "Vic",
    "Wendy",
    "Xander",
    "Yara",
    "Zane",
]

LAST_NAMES = [
    "Smith",
    "Johnson",
    "Williams",
    "Brown",
    "Jones",
    "Garcia",
    "Miller",
    "Davis",
    "Rodriguez",
    "Martinez",
    "Hernandez",
    "Lopez",
    "Gonzalez",
    "Wilson",
    "Anderson",
    "Thomas",
    "Taylor",
    "Moore",
    "Jackson",
    "Martin",
]

PRODUCT_NAMES = [
    "Wireless Mouse",
    "Mechanical Keyboard",
    "USB-C Hub",
    "Monitor Stand",
    "Webcam HD",
    "Desk Lamp",
    "Laptop Stand",
    "Cable Organizer",
    "Noise-Cancelling Headphones",
    "Bluetooth Speaker",
    "External SSD",
    "Portable Charger",
    "Smart Watch",
    "Fitness Tracker",
    "VR Headset",
    "Drone Mini",
    "Action Camera",
    "E-Reader",
    "Tablet Stylus",
    "Docking Station",
    "Ring Light",
    "Microphone USB",
    "Game Controller",
    "Mouse Pad XL",
    "Screen Protector",
    "Phone Mount",
    "Car Charger",
    "Wall Adapter",
    "HDMI Cable",
    "Ethernet Adapter",
]

CATEGORIES = [
    "Electronics",
    "Accessories",
    "Audio",
    "Computing",
    "Gaming",
    "Photography",
    "Wearables",
    "Storage",
]

ORDER_STATUSES = [
    "pending",
    "processing",
    "shipped",
    "delivered",
    "cancelled",
    "refunded",
]

STREET_NAMES = [
    "Main St",
    "Oak Ave",
    "Cedar Ln",
    "Elm Dr",
    "Pine Rd",
    "Maple Ct",
    "Birch Blvd",
    "Walnut Way",
    "Spruce Pl",
    "Ash Cir",
]

CITIES = [
    "San Francisco",
    "New York",
    "Chicago",
    "Austin",
    "Seattle",
    "Denver",
    "Boston",
    "Portland",
    "Miami",
    "Atlanta",
]

STATES = ["CA", "NY", "IL", "TX", "WA", "CO", "MA", "OR", "FL", "GA"]

TAGS = [
    "premium",
    "new",
    "verified",
    "vip",
    "beta-tester",
    "newsletter",
    "referral",
    "loyalty",
    "early-adopter",
    "enterprise",
]


def _seeded_random(seed: int):
    state = seed

    def _next() -> float:
        nonlocal state
        state = (state * 1664525 + 1013904223) & 0x7FFFFFFF
        return state / 0x7FFFFFFF

    return _next


def _pick(arr: list, rand) -> object:
    return arr[math.floor(rand() * len(arr))]


def _pick_n(arr: list, n: int, rand) -> list:
    shuffled = sorted(arr, key=lambda _: rand())
    return shuffled[:n]


def _generate_digest(rand) -> str:
    hex_chars = "0123456789abcdef"
    return "".join(hex_chars[math.floor(rand() * 16)] for _ in range(40))


# ---------------------------------------------------------------------------
# User records (conn-1, test:users)
# ---------------------------------------------------------------------------


def _generate_user_records() -> list[AerospikeRecord]:
    rand = _seeded_random(42)
    records: list[AerospikeRecord] = []
    now_ms = int(time.time() * 1000)

    for i in range(1, 51):
        first = _pick(FIRST_NAMES, rand)
        last = _pick(LAST_NAMES, rand)
        email = f"{first.lower()}.{last.lower()}{i}@example.com"
        age = 18 + math.floor(rand() * 52)
        active = rand() > 0.2
        user_tags = _pick_n(TAGS, 1 + math.floor(rand() * 4), rand)

        metadata = {
            "signupDate": f"2025-{str(1 + math.floor(rand() * 12)).zfill(2)}-{str(1 + math.floor(rand() * 28)).zfill(2)}",
            "lastLogin": f"2026-02-{str(1 + math.floor(rand() * 14)).zfill(2)}T{str(math.floor(rand() * 24)).zfill(2)}:{str(math.floor(rand() * 60)).zfill(2)}:00Z",
            "loginCount": math.floor(rand() * 500),
            "preferredLanguage": _pick(["en", "es", "fr", "de", "ja"], rand),
            "timezone": _pick(["US/Pacific", "US/Eastern", "UTC", "Europe/London", "Asia/Tokyo"], rand),
        }

        records.append(
            AerospikeRecord(
                key=RecordKey(
                    namespace="test",
                    set="users",
                    pk=f"user-{str(i).zfill(4)}",
                    digest=_generate_digest(rand),
                ),
                meta=RecordMeta(
                    generation=1 + math.floor(rand() * 10),
                    ttl=-1,
                    lastUpdateMs=now_ms - math.floor(rand() * 86400000 * 30),
                ),
                bins={
                    "name": f"{first} {last}",
                    "email": email,
                    "age": age,
                    "active": active,
                    "metadata": metadata,
                    "tags": user_tags,
                },
            )
        )
    return records


# ---------------------------------------------------------------------------
# Product records (conn-1, test:products)
# ---------------------------------------------------------------------------


def _generate_product_records() -> list[AerospikeRecord]:
    rand = _seeded_random(137)
    records: list[AerospikeRecord] = []
    now_ms = int(time.time() * 1000)

    for i in range(1, 31):
        name = PRODUCT_NAMES[(i - 1) % len(PRODUCT_NAMES)]
        price = round((5 + rand() * 495) * 100) / 100
        category = _pick(CATEGORIES, rand)
        in_stock = rand() > 0.15

        specs = {
            "weight": f"{round(50 + rand() * 2000)}g",
            "dimensions": f"{round(5 + rand() * 30)}x{round(5 + rand() * 20)}x{round(2 + rand() * 10)}cm",
            "color": _pick(["Black", "White", "Silver", "Blue", "Red", "Gray"], rand),
            "warranty": _pick(["6 months", "1 year", "2 years", "3 years"], rand),
            "manufacturer": _pick(["TechCorp", "GadgetPro", "DigitalWorks", "NexGen", "InnoTech"], rand),
            "rating": round((3 + rand() * 2) * 10) / 10,
            "reviewCount": math.floor(rand() * 1200),
        }

        records.append(
            AerospikeRecord(
                key=RecordKey(
                    namespace="test",
                    set="products",
                    pk=f"prod-{str(i).zfill(4)}",
                    digest=_generate_digest(rand),
                ),
                meta=RecordMeta(
                    generation=1 + math.floor(rand() * 5),
                    ttl=-1,
                    lastUpdateMs=now_ms - math.floor(rand() * 86400000 * 60),
                ),
                bins={
                    "name": name,
                    "price": price,
                    "category": category,
                    "inStock": in_stock,
                    "specs": specs,
                },
            )
        )
    return records


# ---------------------------------------------------------------------------
# Order records (conn-2, staging:orders)
# ---------------------------------------------------------------------------


def _generate_order_records() -> list[AerospikeRecord]:
    rand = _seeded_random(256)
    records: list[AerospikeRecord] = []
    now_ms = int(time.time() * 1000)

    for i in range(1, 201):
        order_id = f"ORD-{10000 + i}"
        user_id = f"user-{str(1 + math.floor(rand() * 50)).zfill(4)}"
        status = _pick(ORDER_STATUSES, rand)

        item_count = 1 + math.floor(rand() * 5)
        items: list[dict] = []
        total = 0.0
        for _j in range(item_count):
            item_price = round((10 + rand() * 200) * 100) / 100
            qty = 1 + math.floor(rand() * 3)
            total += item_price * qty
            items.append(
                {
                    "productId": f"prod-{str(1 + math.floor(rand() * 30)).zfill(4)}",
                    "name": _pick(PRODUCT_NAMES, rand),
                    "price": item_price,
                    "quantity": qty,
                }
            )
        total = round(total * 100) / 100

        state_idx = math.floor(rand() * len(STATES))
        address = {
            "street": f"{100 + math.floor(rand() * 9900)} {_pick(STREET_NAMES, rand)}",
            "city": CITIES[state_idx],
            "state": STATES[state_idx],
            "zip": str(10000 + math.floor(rand() * 89999)),
            "country": "US",
        }

        lat = round((25 + rand() * 23) * 1000000) / 1000000
        lng = round((-124 + rand() * 53) * 1000000) / 1000000
        location = {"type": "Point", "coordinates": [lng, lat]}

        records.append(
            AerospikeRecord(
                key=RecordKey(
                    namespace="staging",
                    set="orders",
                    pk=order_id,
                    digest=_generate_digest(rand),
                ),
                meta=RecordMeta(
                    generation=1 + math.floor(rand() * 3),
                    ttl=86400 * 90,
                    lastUpdateMs=now_ms - math.floor(rand() * 86400000 * 30),
                ),
                bins={
                    "orderId": order_id,
                    "userId": user_id,
                    "total": total,
                    "status": status,
                    "items": items,
                    "address": address,
                    "location": location,
                },
            )
        )
    return records


# ---------------------------------------------------------------------------
# Export
# ---------------------------------------------------------------------------

mock_records: dict[str, dict[str, list[AerospikeRecord]]] = {
    "conn-1": {
        "test:users": _generate_user_records(),
        "test:products": _generate_product_records(),
    },
    "conn-2": {
        "staging:orders": _generate_order_records(),
    },
}
